src/GuzzleHttp/ClientSerializerAdapter.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+/*
+ * This file is part of the CsaGuzzleBundle package
+ *
+ * (c) Charles Sarrazin <charles@sarraz.in>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code
+ */
+
+namespace Csa\Bundle\GuzzleBundle\GuzzleHttp;
+
+use GuzzleHttp\Client;
+use GuzzleHttp\Psr7;
+use Psr\Http\Message\RequestInterface;
+
+class ClientSerializerAdapter extends Client
+{
+    /**
+     * @inherit
+     */
+    public function sendAsync(RequestInterface $request, array $options = [])
+    {
+        $options = $this->prepareDefaults($options);
+        $response = parent::sendAsync($request, $options)->wait();
+
+        return $this->processResponse($response);
+    }
+
+    /**
+     * @inherit
+     */
+    public function requestAsync($method, $uri = null, array $options = [])
+    {
+        $options = $this->prepareDefaults($options);
+        $response = parent::requestAsync($method, $uri, $options)->wait();
+
+        return $this->processResponse($response);
+    }
+
+    /**
+     * Hook into the defaults options that cannot be changed by middleware.
+     *
+     * @param array $options
+     *
+     * @return array
+     */
+    private function prepareDefaults(array $options = [])
+    {
+        $serializerAdapter = $this->getConfig('serializer_adapter');
+
+        if ((isset($options['json'])) && $serializerAdapter) {
+            $modify['body'] = Psr7\stream_for($serializerAdapter->serialize($options['json'], 'json', []));
+            $options['_conditional']['Content-Type'] = 'application/json';
+            unset($options['json']);
+        }
+
+        if ((isset($options['xml'])) && $serializerAdapter) {
+            $modify['body'] = Psr7\stream_for($serializerAdapter->serialize($options['xml'], 'xml', []));
+            $options['_conditional']['Content-Type'] = 'application/xml';
+            unset($options['xml']);
+        }
+
+        return $options;
+    }
+
+    /**
+     * Try to deserialize the response.
+     *
+     * @param Psr7\Response $response
+     *
+     * @return mixed
+     */
+    private function processResponse(Psr7\Response $response)
+    {
+        $serializerAdapter = $this->getConfig('serializer_adapter');
+
+        if ($response->hasHeader('Content-Type') && $serializerAdapter) {
+            $format = false;
+            if (in_array('application/json', $response->getHeader('Content-Type'))) {
```

Comment:
For this part, we should instead check whether the header contains: `application/*+json`, `application/json` or `text/javascript`.

Same thing for XML below: `application/*+xml`, `text/xml`, `application/xml`.